crates/ethrpc/src/current_block/mod.rs

```diff
@@ -16,6 +16,130 @@ use {
+
+impl CurrentBlockStream {
+    /// Returns a reference to the current block. Keeping this reference
+    /// alive locks a mutex so it should be used as shortly as possible.
+    pub fn current(&self) -> watch::Ref<'_, BlockInfo> {
+        self.0.watch_receiver.borrow()
+    }
+
+    /// Returns a stream that always yields the most recent block (if unseen).
+    /// Also yields the current block right away and not just the next new
+    /// block. Use this if you only need to know about the current state.
+    pub fn watch_stream(&self) -> WatchStream<BlockInfo> {
+        WatchStream::new(self.0.watch_receiver.clone())
+    }
+
+    /// Returns a stream that yields every new block **AFTER** it got created.
+    /// Use this if it's required to not miss any blocks.
+    pub fn buffering_stream(&self) -> impl Stream<Item = BlockInfo> {
+        let receiver = self.0.broadcast_sender.subscribe();
+        BroadcastStream::new(receiver).map(Result::unwrap)
+    }
+
+    /// Spawns a new background task that watches for new blocks and notifies
+    /// the [`CurrentBlockStream`] about them.
+    pub async fn new(url: Url, poll_interval: Duration) -> Result<Self> {
+        // Build new Web3 specifically for the current block stream to avoid batching
+        // requests together on chains with a very high block frequency.
+        let web3 = Web3::new(Web3Transport::new(HttpTransport::new(
+            Default::default(),
+            url,
+            "block_stream".into(),
+        )));
+        let web3 = instrument_with_label(&web3, "base_currentBlockStream".into());
+        let first_block = web3.current_block().await?;
+        tracing::debug!(number=%first_block.number, hash=?first_block.hash, "polled block");
+
+        let stream = CurrentBlockStream(Arc::new(BlockStreamInner::new(first_block)));
+        let stream_clone = stream.clone();
+
+        let update_future = async move {
+            let mut previous_block = first_block;
+            loop {
+                tokio::time::sleep(poll_interval).await;
+                let block = match web3.current_block().await {
+                    Ok(block) => block,
+                    Err(err) => {
+                        tracing::warn!("failed to get current block: {:?}", err);
+                        continue;
+                    }
+                };
+
+                // If the block is exactly the same, ignore it.
+                if previous_block.hash == block.hash {
+                    continue;
+                }
+
+                // The new block is different but might still have the same number.
+                tracing::debug!(number=%block.number, hash=?block.hash, "polled block");
+                update_block_metrics(previous_block.number, block.number);
+
+                // Only update the stream if the number has increased.
+                if block.number <= previous_block.number {
+                    continue;
+                }
```

Comment:
I see this is where #2740 comes into play? I think we can add logic here where if the number is > previous_block.number + 1, we specifically poll the blocks from previous_block.number +1 until this one to make the stream gap free (should be a separate PR/task though as it seems somewhat involved if we also want to handle reorgs).

Reply:
Oh, right. Simply looping over all the blocks we know we missed would be way simpler than introducing web socket support. 👍